Add xilinx network support

mchindri opened the pull request:
Created to debug a new tinyiiod version.
 - For tinyiiod seems to work, but it doesn't has the same performance as in linux.
 - Some file are token from here: https://github.com/Xilinx/embeddedsw/tree/master/lib/sw_apps/lwip_echo_server/src
   because it seems that there is support for multiple xilinx boards.
 - Tested on zc706 and zedboard
 - At the moment lwip must be enabled manually from BSP with LWIP_DHCP set to true. (I failed to add them from makefile but there should be a way. I didn't spend too much time on that )

buha commented:
So I just had a look at your branch - you seem to place the xilinx ethernet files here:
```
$ ls ../../network/xil_eth/
i2c_access.c  iic_phyreset.c  platform.c  platform_config.h  platform.h  platform_mb.c  platform_ppc.c  platform_zynq.c  platform_zynqmp.c  sfp.c  si5324.c  xil_eth.c  xil_eth.h
```

By modifying util.tcl like this:
```
-                -template  {Empty Application}
+                -template  {lwIP Echo Server}
```

And runing make on your branch, I get these files:
```
$ ls build/app/src/
echo.c  i2c_access.c  iic_phyreset.c  lscript.ld  main.c  platform.c  platform_config.h  platform.h  platform_mb.c  platform_ppc.c  platform_zynq.c  platform_zynqmp.c  README.txt  sfp.c  si5324.c  Xilinx.spec
```
This shows that the first commit can be dropped, just a slight adaptation of the build system:
 - util.tcl change above + a way to select the template from the Makefile (new parameter for util.tcl?)
 - src.mk get the etnernet files from the autogenerated location
 - remove that build/app/src/main.c (if files are included with SRC_DIRS += ..., otherwise not needed) 
 - edit the build/app/src/platform_config.h file with a script to automatically add a #include "app_config.h" or whatever is needed.
 - ... (maybe other changes)

mchindri commented:
v2
 - Fix bug (it seems to work ok now)
 - Won't try at this moment to solve the project creation. I will focus on iiod and trigger for the moment.

Commit:
network:xil_eth:Implement network functions

Implement network function for network_interface
that are needed for tinyiiod to work.

Signed-off-by: Mihail Chindris <mihail.chindris@analog.com>

Commit:
network:xil_eth: Add common xilinx ethernet files

These files are taken from
https://github.com/Xilinx/embeddedsw/tree/master/lib/sw_apps/lwip_echo_server/src

They will be used to create in order to integrate ethernet in No-OS
stack

Signed-off-by: Mihail Chindris <mihail.chindris@analog.com>

Commit:
iio:iio_app: Add xil eth support

Some header reorder was needed, some are just visual
improvements.
In linux.mk USE_TCP_SOCKET is added by default when
TINYIIOD is set to y to make things easier in iio_app

Signed-off-by: Mihail Chindris <mihail.chindris@analog.com>

buha commented on network/xil_eth/xil_eth.c:
```diff
@@ -0,0 +1,654 @@
+/***************************************************************************//**
+ * @file xil_eth.c
+ * @brief Implementation of ethernet functions for network.
+ * @author Mihail Chindris (mihail.chindrs@analog.com)
```
some typos in comments:
- your name specifically
- LINUX_SOCKET_H_ i guess is a remnant from a template

buha commented:
> * Won't try at this moment to solve the project creation. I will focus on iiod and trigger for the moment.

okay, makes sense, we can do that later

mchindri commented:
v3: 
 - Fix comments
 - Add eth support for xilinx in iio_app (maybe make sens an other PR for it but for the moment it may be ok here)

adrimbarean commented on network/xil_eth/xil_eth.c:
```diff
@@ -0,0 +1,654 @@
+		 * Should some locking be use here?
+		 */
+		/* Concatenate new buffer to p */
+		pbuf_cat(sock->p, p);
```
Question: do we need to update `sock->p_idx` here or does `pbuf_cat` update it itself?

mchindri replied:
I assume p is put at the end of the list. 
p_idx is the index in the buffer of the first element

adrimbarean commented on network/xil_eth/xil_eth.c:
```diff
@@ -0,0 +1,654 @@
+	}
+
+	for (i = 0; i < MAX_SOCKETS; ++i) {
+		cli_sock = &desc->sockets[i];
```
nitpick: why not `_get_sock` here as well?

adrimbarean commented on iio/iio_app/iio_app.c:
```diff
@@ -47,36 +47,53 @@
 #include "no-os/delay.h"
 #include <stdio.h>
 
+#undef USE_TCP_SOCKET
```
Question: what is the point of undefining `USE_TCP_SOCKET` here?

mchindri replied:
I forgot to remove it. It was used just for testing

buha commented on network/xil_eth/xil_eth.c:
```diff
@@ -0,0 +1,654 @@
+		dhcp_start(netif);
+		dhcp_timoutcntr = 24;
+
+		while(((netif->ip_addr.addr) == 0) && (dhcp_timoutcntr > 0))
```
did you mean `dhcp_timoutcntr--`?

buha commented on network/xil_eth/xil_eth.c:
```diff
@@ -0,0 +1,654 @@
+		if (dhcp_timoutcntr <= 0) {
+			if ((netif->ip_addr.addr) == 0) {
+				pr_err("DHCP Timeout\r\n");
+				ret = FAILURE;
```
I think there is an -ETIMEDOUT

buha commented on network/xil_eth/xil_eth.c:
```diff
@@ -0,0 +1,654 @@
+	if (!xemac_add(netif, &ipaddr, &netmask, &gw,
+		       param->mac_ethernet_address, param->emac_baseaddr)) {
+		pr_err("Error adding N/W interface\n\r");
+		ret = FAILURE;
```
also here, a more meaningful error code

buha commented on network/xil_eth/xil_eth.c:
```diff
@@ -0,0 +1,654 @@
+int32_t xil_eth_get_network_interface(struct xil_eth_desc *desc,
+				      struct network_interface **net)
+{
+	*net = &desc->noos_net;
```
this dereferencing could fail
